src/sas/qtgui/Calculators/Shape2SAS/Constraints.py

```diff
@@ -55,75 +53,297 @@ def __init__(self, parent=None):
+                lhs_shape_num = lhs_base[3:]
+                rhs_shape_num = rhs_base[3:]
+
+                for axis in ['X', 'Y', 'Z']:
+                    lhs_full = f"{'d' if lhs_is_delta else ''}COM{axis}{lhs_shape_num}"
+                    rhs_full = f"{'d' if rhs_is_delta else ''}COM{axis}{rhs_shape_num}"
+                    new_targets.append(ast.Name(id=lhs_full, ctx=ast.Store()))
+                    new_values.append(ast.Name(id=rhs_full, ctx=ast.Load()))
 
-        #Get and check translation
-        translation, checkedPars = self.getTranslation(constraintsStr, importStatement, modelPars, modelVals, checkedPars)
+            else:
+                # rhs is a regular parameter: COM2 = X -> COMX2, COMY2, COMZ2 = X, X, X
+                lhs_shape_num = lhs_base[3:]
+                rhs_full = f"{'d' if rhs_is_delta else ''}{rhs_base}"
+                for axis in ['X', 'Y', 'Z']:
+                    lhs_full = f"{'d' if lhs_is_delta else ''}COM{axis}{lhs_shape_num}"
+                    new_targets.append(ast.Name(id=lhs_full, ctx=ast.Store()))
+                    new_values.append(ast.Name(id=rhs_full, ctx=ast.Load()))
+
+            constraint.targets = [ast.Tuple(elts=new_targets, ctx=ast.Store())]
+            constraint.value = ast.Tuple(elts=new_values, ctx=ast.Load())
+            return constraint
+
+        def parse_ast(tree: ast.AST):
+            params = None
+            imports = []
+            constraints = []
 
-        return importStatement, parameters, translation, checkedPars
+            for node in ast.walk(tree):
+                match node:
+                    case ast.ImportFrom() | ast.Import():
+                        imports.append(node)
+
+                    case ast.Assign():
+                        if len(node.targets) != 1 or isinstance(node.targets[0], ast.Tuple) or isinstance(node.value, ast.Tuple):
+                            self.log_embedded_error(f"Tuple assignment is not supported (line {node.lineno}).")
+                            raise ValueError(f"Tuple assignment is not supported (line {node.lineno}).")
+
+                        if node.targets[0].id == 'parameters':
+                            params = node
+                            continue
+
+                        if node.targets[0].id.startswith('dCOM') or node.targets[0].id.startswith('COM'):
+                            constraints.append(expand_center_of_mass_pars(node))
+                        else:
+                            constraints.append(node)
+            return params, imports, constraints
+
+        def extract_symbols(constraints: list[ast.AST]) -> tuple[list[str], list[str]]:
+            """Extract all symbols used in the constraints."""
+            lhs, rhs = set(), set()
+            lineno = {}
+            for node in constraints:
+                # left-hand side of assignment
+                for target in node.targets:
```

Comment:
The match in `node.targets` and `ast.walk` loops is exactly the same. Either find a way to combine the two loops, put the match/case into its own function/method, or both.